lib/sdam/server.js

```diff
@@ -1,30 +1,109 @@
 'use strict';
 const EventEmitter = require('events');
+const MongoError = require('../error').MongoError;
+const Pool = require('../connection/pool');
+const relayEvents = require('../utils').relayEvents;
+const calculateDurationInMs = require('../utils').calculateDurationInMs;
+const Query = require('../connection/commands').Query;
+const TwoSixWireProtocolSupport = require('../wireprotocol/2_6_support');
+const ThreeTwoWireProtocolSupport = require('../wireprotocol/3_2_support');
+const BSON = require('../connection/utils').retrieveBSON();
+const createClientInfo = require('../topologies/shared').createClientInfo;
+const Logger = require('../connection/logger');
+const ServerDescription = require('./server_description').ServerDescription;
+const ReadPreference = require('../topologies/read_preference');
 
+/**
+ *
+ * @fires Server#serverHeartbeatStarted
+ * @fires Server#serverHeartbeatSucceeded
+ * @fires Server#serverHeartbeatFailed
+ */
 class Server extends EventEmitter {
-  constructor(description) {
+  /**
+   * Create a server
+   *
+   * @param {ServerDescription} description
+   * @param {Object} options
+   */
+  constructor(description, options) {
     super();
 
     this.s = {
-      description
+      // the server description
+      description,
+      // a saved copy of the incoming options
+      options,
+      // the server logger
+      logger: Logger('Server', options),
+      // the bson parser
+      bson:
+        options.bson ||
+        new BSON([
+          BSON.Binary,
+          BSON.Code,
+          BSON.DBRef,
+          BSON.Decimal128,
+          BSON.Double,
+          BSON.Int32,
+          BSON.Long,
+          BSON.Map,
+          BSON.MaxKey,
+          BSON.MinKey,
+          BSON.ObjectId,
+          BSON.BSONRegExp,
+          BSON.Symbol,
+          BSON.Timestamp
+        ]),
+      // client metadata for the initial handshake
+      clientInfo: createClientInfo(options),
     };
   }
 
   get description() {
     return this.s.description;
   }
 
+  get name() {
+    return this.s.description.address;
+  }
+
   /**
    * Initiate server connect
    *
    * @param {Array} [options.auth] Array of auth options to apply on connect
    */
-  connect(options, callback) {
+  connect(options) {
```

Comment:
I had a long think about this one! When would you call back from connect? When all pool connections are connected? When the first one is connected? I decided to handle this through events instead of callbacks